README.md
## Installation
```bash
composer require sevenecks/ansi
```

## Colorize

### Usage

#### Static Usage
The methods on Colorize can be called statically and the full API is listed later. 

```php
require_once __DIR__ . '/vendor/autoload.php';
use SevenEcks\Ansi\Colorize;

$test_string = 'This is a test';

echo Colorize::italic($test_string) . "\n";
echo Colorize::cyan($test_string) . "\n";
echo Colorize::bgRed($test_string) . "\n";
// background color first
echo Colorize::cyan(Colorize::bgRed($test_string)) . "\n";
// manually call the colorize string function
echo Colorize::colorizeString("TEST STRING", Colorize::$FOREGROUND_BLUE);
```

#### Instantiated Usage
Colorize also supports a magic method that will allow you to call the static methods dynamically, if you want to instantiate the Colorize class. This can be useful for dependency injection.

```php
require_once __DIR__ . '/vendor/autoload.php';
use SevenEcks\Ansi\Colorize;

$test_string = 'This is a test';

$colorize = new Colorize;

echo $color->red('tester');
echo $color->darkGray('testy mctesterson');
```

### Colors

There are several colors available for foreground and background.

#### Foreground

Colorize::italic
Colorize::bold
Colorize::underline
Colorize::invert
Colorize::darkGray
Colorize::blue
Colorize::lightBlue
Colorize::green
Colorize::lightGreen
Colorize::cyan
Colorize::lightCyan
Colorize::red
Colorize::lightRed
Colorize::purple
Colorize::lightPurple
Colorize::brown
Colorize::yellow
Colorize::lightGray
Colorize::white

#### Background

Colorize::bgBlack
Colorize::bgRed
Colorize::bgGreen
Colorize::bgYellow
Colorize::bgBlue
Colorize::bgMagenta
Colorize::bgCyan
Colorize::bgLightGray

## API
```php
//foreground colors
public static $FOREGROUND_BLACK = '0;30';
public static $FOREGROUND_DARK_GRAY = '1;30';
public static $FOREGROUND_BLUE = '0;34';
public static $FOREGROUND_LIGHT_BLUE = '1;34';
public static $FOREGROUND_GREEN = '0;32';
public static $FOREGROUND_LIGHT_GREEN = '1;32';
public static $FOREGROUND_CYAN = '0;36';
public static $FOREGROUND_LIGHT_CYAN = '1;36';
public static $FOREGROUND_RED = '0;31';
public static $FOREGROUND_LIGHT_RED = '1;31';
public static $FOREGROUND_PURPLE = '0;35';
public static $FOREGROUND_LIGHT_PURPLE = '1;35';
public static $FOREGROUND_BROWN = '0;33';
public static $FOREGROUND_YELLOW = '1;33';
public static $FOREGROUND_LIGHT_GRAY = '0;37';
public static $FOREGROUND_WHITE = '1;37';
public static $FOREGROUND_BOLD = '1';
public static $FOREGROUND_UNDERLINE = '4';
public static $FOREGROUND_ITALIC = '3'; 
public static $FOREGROUND_INVERT = '7';
public static $FOREGROUND_STRIKETHROUGH = '9';

// background colors 
public static $BACKGROUND_BLACK = '40';
public static $BACKGROUND_RED = '41';
public static $BACKGROUND_GREEN = '42';
public static $BACKGROUND_YELLOW = '43';
public static $BACKGROUND_BLUE = '44';
public static $BACKGROUND_MAGENTA = '45';
public static $BACKGROUND_CYAN = '46';
public static $BACKGROUND_LIGHT_GRAY = '47'; 

/**
 * Dynamically call the static methods from instance variables
 * and provide the correct arguments as needed
 *
 * @param string $name
 * @param array $arguments
 * @return string
 */
public function __call($name, $arguments)

/**
* Colorize a string of text and return it
*
* @param string $text
* @param string $color
*
* @return string
*/
public static function colorizeString($text, $color) 

/**
 * strikethrough some text 
 *
 * @param string $text
 * @return string
 */
public static function strikethrough($text)

/**
 * invert the foreground of a string
 *
 * @param string $text
 * @return string
 */
public static function invert($text)

/**
 * bold the foreground of a string
 *
 * @param string $text
 * @return string
 */
public static function bold($text)

/**
 * Italic the foreground of a string
 *
 * @param string $text
 * @return string
 */
public static function italic($text)

/**
 * Underline the foreground of a string
 *
 * @param string $text
 * @return string
 */
public static function underline($text)

/**
* Colorize the foreground of a string
*
* @param string $text
* @return string
*/
public static function black($text)
public static function darkGray($text)
public static function blue($text)
public static function lightBlue($text)
public static function green($text)
public static function lightGreen($text)
public static function cyan($text)
public static function lightCyan($text)
public static function red($text)
public static function lightRed($text)
public static function purple($text)
public static function lightPurple($text)
public static function brown($text)
public static function yellow($text)
public static function lightGray($text)
public static function white($text)

/**
 * Colorize the background of a string
 *
 * @param string $text
 * @return string
 */
public static function bgBlack($text)
public static function bgRed($text)
public static function bgGreen($text)
public static function bgYellow($text)
public static function bgBlue($text)
public static function bgMagenta($text)
public static function bgCyan($text)
public static function bgLightGray($text)

## ColorInterface

You can use the ColorInterface.php file to implement your own ANSI Colorization class. The interface requires all the functions that the Colorize class implements, but does not care about how you define your colors or what is going on below the hood.

```
## Change Log
Please see [Change Log](CHANGELOG.md) for more information.

## License
The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
## 1.0.8 - 2018-09-11 2:48PM

- added strikethrough (seems to not work on iTerm2)

## 1.0.7 - 2018-08-23 4:47PM

- Added invert(), bold(), italic(), and underline() methods

## 1.0.6 - 2018-03-20 3:49PM

- Removed tests since they were unused

## 1.0.5 - 2018-03-16 5:13PM

- Added ability ot instantiate class and dynamically call any static methods
- Added ColorInterface which Colorize implements
- Changes make the class sutiable for dependency injection
- Updated composer dependencies to remove PHPUnit since it wasn't being used

## 1.0.4 - 2018-03-16 5:08PM

- Updated README.md with new methods

## 1.0.3 - 2017-12-12 12:24PM

- Updated README.md with list of available colors and example of manually colorizing a string

## 1.0.2 - 2017-12-12 11:54PM

- Updated README.md

## 1.0.1 - 2017-12-11 1:40PM

- Updated ReadMe examples

## 1.0.0 - 2017-12-08 1:47PM

- Initial release

## 0.1.1 - 2017-12-08 1:43PM

- Updated readme
- Updated composer.json

## 0.1.0 - 2017-12-08 12:30PM

- First commits


tests/ColorizeTest.php
<?php

namespace SevenEcks\Ansi\Test;

use SevenEcks\Ansi\Colorize;

class ColorizeTest extends \PHPUnit_Framework_TestCase
{
    /** @test */
    public function it_has_colors()
    {
        $this->assertEquals('0;30', Colorze::$FOREGROUND_BLACK);
    }
}


src/Colorize.php
<?php
namespace SevenEcks\Ansi;

use SevenEcks\Ansi\ColorInterface;

/**
 * Utilities for dealing with ansi within PHP. Useful for colorizing terminal output.
 *
 * @author Brendan Butts <bbutts@stormcode.net>
 */
class Colorize implements ColorInterface {
    //foreground colors
    public static $FOREGROUND_BLACK = '0;30';
    public static $FOREGROUND_DARK_GRAY = '1;30';
    public static $FOREGROUND_BLUE = '0;34';
    public static $FOREGROUND_LIGHT_BLUE = '1;34';
    public static $FOREGROUND_GREEN = '0;32';
    public static $FOREGROUND_LIGHT_GREEN = '1;32';
    public static $FOREGROUND_CYAN = '0;36';
    public static $FOREGROUND_LIGHT_CYAN = '1;36';
    public static $FOREGROUND_RED = '0;31';
    public static $FOREGROUND_LIGHT_RED = '1;31';
    public static $FOREGROUND_PURPLE = '0;35';
    public static $FOREGROUND_LIGHT_PURPLE = '1;35';
    public static $FOREGROUND_BROWN = '0;33';
    public static $FOREGROUND_YELLOW = '1;33';
    public static $FOREGROUND_LIGHT_GRAY = '0;37';
    public static $FOREGROUND_WHITE = '1;37';
    public static $FOREGROUND_BOLD = '1';
    public static $FOREGROUND_UNDERLINE = '4';
    public static $FOREGROUND_ITALIC = '3'; 
    public static $FOREGROUND_INVERT = '7';
    public static $FOREGROUND_STRIKETHROUGH = '9';

    // background colors 
    public static $BACKGROUND_BLACK = '40';
    public static $BACKGROUND_RED = '41';
    public static $BACKGROUND_GREEN = '42';
    public static $BACKGROUND_YELLOW = '43';
    public static $BACKGROUND_BLUE = '44';
    public static $BACKGROUND_MAGENTA = '45';
    public static $BACKGROUND_CYAN = '46';
    public static $BACKGROUND_LIGHT_GRAY = '47'; 

    /**
     * Dynamically call the static methods from instance variables
     * and provide the correct arguments as needed
     *
     * @TODO make this check if the function actually exists
     *
     * @param string $name
     * @param array $arguments
     * @return string
     */
    public function __call($name, $arguments)
    {
        return self::$name(...$arguments);
    }

    /**
     * Colorize a string of text and return it
     *
     * @param string $text
     * @param string $color
     *
     * @return string
     */
    public static function colorizeString($text, $color) 
    {
        return "\033[" . $color . "m" . $text . "\033[0m";
    }

    /**
     * invert the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function invert($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_INVERT);
    }

    /**
     * strikethrough some text 
     *
     * @param string $text
     * @return string
     */
    public static function strikethrough($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_STRIKETHROUGH);
    }

    /**
     * bold the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function bold($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_BOLD); 
    }

    /**
     * Italic the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function italic($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_ITALIC); 
    }

    /**
     * Underline the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function underline($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_UNDERLINE); 
    }

    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function black($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_BLACK); 
    }

    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function darkGray($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_DARK_GRAY); 
    }

    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function blue($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_BLUE); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightBlue($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_BLUE); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function green($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_GREEN); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightGreen($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_GREEN); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function cyan($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_CYAN); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightCyan($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_CYAN); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function red($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_RED); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightRed($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_RED); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function purple($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_PURPLE); 
    } 

    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightPurple($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_PURPLE); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function brown($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_BROWN); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function yellow($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_YELLOW); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function lightGray($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_LIGHT_GRAY); 
    }
    
    /**
     * Colorize the foreground of a string
     *
     * @param string $text
     * @return string
     */
    public static function white($text)
    {
       return self::colorizeString($text, self::$FOREGROUND_WHITE); 
    }

    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgBlack($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_BLACK); 
    }

    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgRed($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_RED); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgGreen($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_GREEN); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgYellow($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_YELLOW); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgBlue($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_BLUE); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgMagenta($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_MAGENTA); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgCyan($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_CYAN); 
    }
    
    /**
     * Colorize the background of a string
     *
     * @param string $text
     * @return string
     */
    public static function bgLightGray($text)
    {
       return self::colorizeString($text, self::$BACKGROUND_LIGHT_GRAY); 
    }
}

?>
